modules/plugin/svg/src/main/java/org/geotools/renderer/style/RenderableSVGCache.java

```diff
@@ -0,0 +1,260 @@
+import org.opengis.feature.Feature;
+import org.opengis.filter.expression.Expression;
+import org.w3c.dom.Document;
+import org.w3c.dom.Element;
+import org.w3c.dom.NamedNodeMap;
+import org.w3c.dom.Node;
+import org.w3c.dom.NodeList;
+import org.xml.sax.EntityResolver;
+import org.xml.sax.InputSource;
+import org.xml.sax.SAXException;
+
+/** Cache for RenderableSVG instances */
+public class RenderableSVGCache {
+
+    private static final Pattern PARAMETER_PATTERN = Pattern.compile("param\\((.+)\\).*");
+
+    /** Parsed SVG glyphs cache */
+    static Map<String, RenderableSVG> glyphCache =
+            Collections.synchronizedMap(new SoftValueHashMap<>());
+
+    /** The possible mime types for SVG */
+    static final Set<String> formats = new HashSet<String>();
+
+    static {
+        formats.add("image/svg");
+        formats.add("image/svg-xml");
+        formats.add("image/svg+xml");
+    }
+
+    private EntityResolver resolver;
+
+    public RenderableSVGCache() {
+        this(null);
+    }
+
+    public RenderableSVGCache(Map<Key, Object> hints) {
+        if (hints != null && hints.containsKey(Hints.ENTITY_RESOLVER)) {
+            // use entity resolver provided (even if null)
+            this.resolver = (EntityResolver) hints.get(Hints.ENTITY_RESOLVER);
+
+            if (this.resolver == null) { // use null instance rather than check each time
+                this.resolver = NullEntityResolver.INSTANCE;
+            }
+        } else {
+            this.resolver = GeoTools.getEntityResolver(null);
+        }
+    }
+
+    public RenderableSVG getRenderableSVG(Feature feature, Expression url, String format)
+            throws Exception {
+        // check we do support the declared format
+        if (format == null || !formats.contains(format.toLowerCase())) return null;
+
+        // grab the url
+        String svgfile = url.evaluate(feature, String.class);
+        if (svgfile == null) {
+            throw new IllegalArgumentException(
+                    "The specified expression could not be turned into an URL");
+        } else {
+            // just for validation parse the URL
+            if (Converters.convert(svgfile, URL.class) == null) {
+                throw new IllegalArgumentException("Invalid URL: " + svgfile);
+            }
+        }
+
+        // turn the svg into a document and cache results
+        RenderableSVG svg = glyphCache.get(svgfile);
+        if (svg == null) {
+            String parser = XMLResourceDescriptor.getXMLParserClassName();
+            SAXSVGDocumentFactory f =
+                    new SAXSVGDocumentFactory(parser) {
+                        @Override
+                        public InputSource resolveEntity(String publicId, String systemId)
+                                throws SAXException {
+                            InputSource source = super.resolveEntity(publicId, systemId);
+                            if (source == null) {
```

Comment:
So basic we give it a first try with the default resolver, and then only if it fails we try the one provided one as an hint ?

Reply:
Hmm... you have a point. This code has not been modified compared to master:
https://github.com/geotools/geotools/blob/master/modules/plugin/svg/src/main/java/org/geotools/renderer/style/SVGGraphicFactory.java#L165

However, if we disallow entity resolution in the one provided in the hints, it should stay that way, that one should be used first.

Reply:
Checked, we actually need the code to stay as is. The "super" resolver only picks from selected sources in the classpath, and the default GeoTools resolver won't allow DTD, making the entire SVG parsing fail.

Reply:
Ah! thanks for the info.